README.md
# thousands

[![Build Status](https://travis-ci.org/tov/thousands-rs.svg?branch=master)](https://travis-ci.org/tov/thousands-rs)
[![Crates.io](https://img.shields.io/crates/v/thousands.svg?maxAge=2592000)](https://crates.io/crates/thousands)
[![License: MIT](https://img.shields.io/badge/license-MIT-blue.svg)](LICENSE-MIT)
[![License: Apache 2.0](https://img.shields.io/badge/license-Apache_2.0-blue.svg)](LICENSE-APACHE)

Provides a trait, `Separable`, for formatting numbers with separators
between the digits. Typically this will be used to add commas or spaces
every three digits from the right, but can be configured via a
`SeparatorPolicy`.

## Examples

The simplest way to use the library is with trait `Separable`’s method
`separate_with_commas` method, which does what it sounds like:

```rust
use thousands::Separable;

 assert_eq!(   12345  .separate_with_commas(),  "12,345" );
 assert_eq!( (-12345) .separate_with_commas(), "-12,345" );
 assert_eq!(    9876.5.separate_with_commas(),   "9,876.5" );
```

There are also methods `separate_with_spaces`, `separate_with_dots`, and
`separate_with_underscores`, in case you, your culture, or your file
format prefer those separators.

However, it's also possible to pass a policy for different behavior:

```rust
use thousands::{Separable, SeparatorPolicy, digits};

let policy = SeparatorPolicy {
    separator: ',',
    groups:    &[3, 2],
    digits:    digits::ASCII_DECIMAL,
};

assert_eq!( 1234567890.separate_by_policy(policy), "1,23,45,67,890" );
```

## Usage

It’s [on crates.io](https://crates.io/crates/thousands), so you can add

```toml
[dependencies]
thousands = "0.1.4"
```

to your `Cargo.toml`.

This crate supports Rust version 1.22 and newer.



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog] and this project adheres to
[Semantic Versioning].

[Keep a Changelog]: http://keepachangelog.com/en/1.0.0/
[Semantic Versioning]: http://semver.org/spec/v2.0.0.html

## [0.1.4] - 2019-10-19

### Changed
- The separator is now a `&str` rather than a `char`.

### Added
- A non-blanket `impl Separable for str` allocates only once, for the result,
rather than twice as the blanket `Display`-based `impl` does.
- An empty group array results in no separators.

### Fixed
- Now respects UTF-8 for the separator as well.

## [0.1.4] - 2019-10-19

### Fixed
- Now respects UTF-8.

## [0.1.3] - 2019-10-19

### Added
- `Separable::separate_with_underscores` method and
`policies::UNDERSCORE_SEPARATOR` constant.

## [0.1.2] - 2018-09-18

### Fixed
- Github URL in Cargo.toml.

## [0.1.1] - 2018-09-16

### Added
- Doc comment for `policies` module.

## [0.1.0] - 2018-09-16

Initial release.



src/traits.rs
use crate::{SeparatorPolicy, policies};

/// Provides methods for formatting numbers with separators between the digits.
pub trait Separable {
    /// Inserts a comma every three digits from the right.
    ///
    /// This is equivalent to `self.separate_by_policy(policies::COMMA_SEPARATOR)`.
    ///
    /// # Examples
    ///
    /// ```
    /// # use thousands::*;
    /// assert_eq!( 12345.separate_with_commas(), "12,345" );
    /// ```
    fn separate_with_commas(&self) -> String {
        self.separate_by_policy(policies::COMMA_SEPARATOR)
    }

    /// Inserts a space every three digits from the right.
    ///
    /// This is equivalent to `self.separate_by_policy(policies::SPACE_SEPARATOR)`.
    ///
    /// # Examples
    ///
    /// ```
    /// # use thousands::*;
    /// assert_eq!( 12345.separate_with_spaces(), "12 345" );
    /// ```
    fn separate_with_spaces(&self) -> String {
        self.separate_by_policy(policies::SPACE_SEPARATOR)
    }

    /// Inserts a period every three digits from the right.
    ///
    /// This is equivalent to `self.separate_by_policy(policies::DOT_SEPARATOR)`.
    ///
    /// # Examples
    ///
    /// ```
    /// # use thousands::*;
    /// assert_eq!( 12345.separate_with_dots(), "12.345" );
    /// ```
    fn separate_with_dots(&self) -> String {
        self.separate_by_policy(policies::DOT_SEPARATOR)
    }

    /// Inserts an underscore every three digits from the right.
    ///
    /// This is equivalent to `self.separate_by_policy(policies::UNDERSCORE_SEPARATOR)`.
    ///
    /// # Examples
    ///
    /// ```
    /// # use thousands::*;
    /// assert_eq!( 12345.separate_with_underscores(), "12_345" );
    /// ```
    fn separate_with_underscores(&self) -> String {
        self.separate_by_policy(policies::UNDERSCORE_SEPARATOR)
    }

    /// Adds separators according to the given [`SeparatorPolicy`].
    ///
    /// # Examples
    ///
    /// ```
    /// use thousands::{Separable, SeparatorPolicy, digits};
    ///
    /// let policy = SeparatorPolicy {
    ///     separator:  ":",
    ///     groups:     &[1, 2, 3, 4],
    ///     digits:     digits::ASCII_DECIMAL,
    /// };
    ///
    /// assert_eq!( 1234567654321u64.separate_by_policy(policy),
    ///             "123:4567:654:32:1" );
    /// ```
    ///
    /// [`SeparatorPolicy`]: struct.SeparatorPolicy.html
    fn separate_by_policy(&self, policy: SeparatorPolicy) -> String;
}



src/display.rs
use std::fmt::Display;

use crate::{Separable, SeparatorPolicy};

impl Separable for str {
    fn separate_by_policy(&self, policy: SeparatorPolicy) -> String {
        let (before, number, after, count) = find_span(&self, |c| policy.digits.contains(&c));
        let formatted = insert_separator_rev(number, policy.separator, policy.groups);

        let mut result = String::with_capacity(before.len() + formatted.len() + after.len());

        result.push_str(before);
        result.extend(formatted.chars().rev());
        result.push_str(after);

        result
    }
}

fn insert_separator_rev(number: &str, sep: char, mut groups: &[u8]) -> String {
    // Does guessing the size like on the next line make sense?
    let mut buffer  = String::with_capacity(2 * number.len());
    let mut counter = 0;

<<<<<<< HEAD
    for c in number.chars().rev() {
        if Some(&counter) == groups.get(0) {
            buffer.push(sep);
            counter = 0;

            if groups.len() > 1 {
                groups = &groups[1 ..];
=======
        for (digit, comma_after) in number.chars().zip(iter) {
            result.push(digit);
            if comma_after {
                result.push_str(policy.separator);
>>>>>>> 2d48dbd... Another UTF-8 fix; handling the empty groups case.
            }
        }

        counter += 1;
        buffer.push(c);
    }

    buffer
}

impl<T: Display> Separable for T {
    fn separate_by_policy(&self, policy: SeparatorPolicy) -> String {
        self.to_string().as_str().separate_by_policy(policy)
    }
}

fn find_span<F>(s: &str, is_digit: F) -> (&str, &str, &str, usize) where F: Fn(char) -> bool {
    let number_start = match s.char_indices()
        .find_map(|(i, c)|
            if is_digit(c) {
                Some(i)
            } else {
                None
            }) {

        Some(len) => len,
        None      => return (s, "", "", 0),
    };

    let mut count = 0;

    let number_end = number_start + match s[number_start ..].char_indices()
        .find_map(|(i, c)|
            if is_digit(c) {
                count += 1;
                None
            } else {
                Some(i)
            }) {

        Some(len) => len,
        None      => s.len() - number_start,
    };

    (&s[.. number_start], &s[number_start .. number_end], &s[number_end ..], count)
}

#[cfg(test)]
mod test {
    use crate::{Separable, SeparatorPolicy, digits, policies};

    #[test]
    fn integer_thousands_commas() {
        assert_eq!( "12345".separate_with_commas(),
                    "12,345" );
    }

    #[test]
    fn smilies() {
        let policy = SeparatorPolicy {
            separator: "😃😃",
            groups:    &[1],
            digits:    &['🙁'],
        };

        assert_eq!( "  🙁🙁🙁🙁🙁  ".separate_by_policy(policy),
                    "  🙁😃😃🙁😃😃🙁😃😃🙁😃😃🙁  " );
    }

    #[test]
    fn three_two_two_two() {
        let policy = SeparatorPolicy {
            separator: ",",
            groups:    &[3, 2],
            digits:    &digits::ASCII_DECIMAL,
        };

        assert_eq!( "1234567890".separate_by_policy(policy),
                    "1,23,45,67,890" );
    }

    #[test]
    fn minus_sign_and_decimal_point() {
        assert_eq!( "-1234.5".separate_with_commas(),
                    "-1,234.5" );
    }

    #[test]
    fn hex_four() {
        assert_eq!( "deadbeef".separate_by_policy(policies::HEX_FOUR),
                    "dead beef" );
    }
}


src/helpers.rs
use crate::SeparatorPolicy;

#[derive(Debug)]
pub struct SeparatorIterator<'a> {
    groups:                  &'a [u8],
    repeat_groups_remaining: usize,
    current_group_index:     usize,
    current_group_size:      usize,
}

impl<'a> SeparatorIterator<'a> {
    pub fn new(policy: &'a SeparatorPolicy, ndigits: usize) -> Self {
        let groups = &policy.groups;

        let mut sum = 0;

        for (index, &group) in groups.into_iter().enumerate() {
            sum += group as usize;

            if ndigits <= sum {
                return SeparatorIterator {
                    groups,
                    repeat_groups_remaining: 0,
                    current_group_index:     index,
                    current_group_size:      ndigits - (sum - group as usize),
                }
            }
        }

        let repeat_group_len = match groups.last() {
            Some(n) => *n as usize,
            None    =>
                return SeparatorIterator {
                    groups:                  &[],
                    repeat_groups_remaining: 0,
                    current_group_index:     0,
                    current_group_size:      0,
                    len,
                }
        };

<<<<<<< HEAD
        let len_remaining = len - sum;
        let (repeat_groups_remaining, current_group_size)
                          = ceil_div_mod(len_remaining, repeat_group_len);
=======
        let digits_remaining = ndigits - sum;
        let (repeat_groups_remaining, current_group_size)
                             = ceil_div_mod(digits_remaining, repeat_group_len);
>>>>>>> Helper for precomputing where commas go works.

        SeparatorIterator {
            groups,
            repeat_groups_remaining,
            current_group_index: groups.len() - 1,
            current_group_size,
        }
    }
}

impl<'a> Iterator for SeparatorIterator<'a> {
    type Item = bool;

    fn next(&mut self) -> Option<Self::Item> {
        self.len = self.len.checked_sub(1)?;

        self.current_group_size = self.current_group_size.saturating_sub(1);
        if self.current_group_size > 0 {
            return Some(false);
        }

        if let Some(repeat_groups_remaining) = self.repeat_groups_remaining.checked_sub(1) {
            self.repeat_groups_remaining = repeat_groups_remaining;
        } else if let Some(current_group_index) = self.current_group_index.checked_sub(1) {
            self.current_group_index = current_group_index;
        } else {
            return Some(false);
        }

        self.current_group_size = self.groups[self.current_group_index] as usize;
        return Some(true);
    }
}

fn ceil_div_mod(n: usize, m: usize) -> (usize, usize) {
    let round_up = n + m - 1;
    (round_up / m, round_up % m + 1)
}

#[cfg(test)]
mod test {
    use crate::*;
    use super::*;

    fn make_policy(groups: &[u8]) -> SeparatorPolicy {
        let mut result = policies::COMMA_SEPARATOR;
        result.groups = groups;
        result
    }

    fn group_string(groups: &[u8], digits: &str) -> String {
        use std::iter::once;

        let policy = &make_policy(groups);
        let iter = SeparatorIterator::new(policy, digits.chars().count());

        digits.chars().zip(iter)
            .flat_map(|(digit, comma_after)|
                    once(digit)
                        .chain(if comma_after { Some(',') } else { None }))
            .collect()
    }

    macro_rules! grouping_test {
        ( $name:ident, $groups:tt, $result:tt ) => {
            #[test]
            fn $name() {
                let result = $result;
                let input = $result.chars().filter(|&c| c != ',').collect::<String>();
                assert_eq!(group_string(&$groups, &input), result);
            }
        };
    }

    grouping_test!(by_nothing_of_0, [], "");
    grouping_test!(by_nothing_of_1, [], "1");
    grouping_test!(by_nothing_of_2, [], "21");
    grouping_test!(by_nothing_of_3, [], "321");

    grouping_test!(by_1s_of_0, [1], "");
    grouping_test!(by_1s_of_1, [1], "1");
    grouping_test!(by_1s_of_2, [1], "2,1");
    grouping_test!(by_1s_of_3, [1], "3,2,1");

    grouping_test!(by_2s_of_0, [2], "");
    grouping_test!(by_2s_of_1, [2], "1");
    grouping_test!(by_2s_of_2, [2], "21");
    grouping_test!(by_2s_of_3, [2], "3,21");
    grouping_test!(by_2s_of_4, [2], "43,21");
    grouping_test!(by_2s_of_5, [2], "5,43,21");
    grouping_test!(by_2s_of_6, [2], "65,43,21");
    grouping_test!(by_2s_of_7, [2], "7,65,43,21");
    grouping_test!(by_2s_of_8, [2], "87,65,43,21");
    grouping_test!(by_2s_of_9, [2], "9,87,65,43,21");

    grouping_test!(by_3s_of_1, [3], "1");
    grouping_test!(by_3s_of_2, [3], "21");
    grouping_test!(by_3s_of_3, [3], "321");
    grouping_test!(by_3s_of_4, [3], "4,321");
    grouping_test!(by_3s_of_5, [3], "54,321");
    grouping_test!(by_3s_of_6, [3], "654,321");
    grouping_test!(by_3s_of_7, [3], "7,654,321");
    grouping_test!(by_3s_of_8, [3], "87,654,321");
    grouping_test!(by_3s_of_9, [3], "987,654,321");

    grouping_test!(by_2s3_of_1, [3, 2], "1");
    grouping_test!(by_2s3_of_2, [3, 2], "21");
    grouping_test!(by_2s3_of_3, [3, 2], "321");
    grouping_test!(by_2s3_of_4, [3, 2], "4,321");
    grouping_test!(by_2s3_of_5, [3, 2], "54,321");
    grouping_test!(by_2s3_of_6, [3, 2], "6,54,321");
    grouping_test!(by_2s3_of_7, [3, 2], "76,54,321");
    grouping_test!(by_2s3_of_8, [3, 2], "8,76,54,321");
    grouping_test!(by_2s3_of_9, [3, 2], "98,76,54,321");

    grouping_test!(by_5s4321_of_20, [1, 2, 3, 4, 5],
                   "KJIHG,FEDCB,A987,654,32,1");
    grouping_test!(by_5s4321_of_16, [1, 2, 3, 4, 5],
                   "G,FEDCB,A987,654,32,1");
    grouping_test!(by_5s4321_of_11, [1, 2, 3, 4, 5],
                   "B,A987,654,32,1");
    grouping_test!(by_5s4321_of_10, [1, 2, 3, 4, 5],
                   "A987,654,32,1");
    grouping_test!(by_5s4321_of_9, [1, 2, 3, 4, 5],
                   "987,654,32,1");
    grouping_test!(by_5s4321_of_7, [1, 2, 3, 4, 5],
                   "7,654,32,1");
    grouping_test!(by_5s4321_of_1, [1, 2, 3, 4, 5],
                   "1");
    grouping_test!(by_5s4321_of_0, [1, 2, 3, 4, 5],
                   "");
}

#[cfg(test)]
mod sep_len_test {
    use super::test_common::*;

    fn run_iterator(mut iter: SeparatorIterator) -> (Vec<usize>, Vec<usize>) {
        let mut predictions = Vec::with_capacity(iter.len());
        let mut actuals     = Vec::with_capacity(iter.len());

        let mut prediction;
        while let Some(actual) = {
            prediction = iter.sep_len();
            iter.next()
        } {
            predictions.push(prediction);
            actuals.push(if actual { 1 } else { 0 })
        }

        let mut acc = 0;
        for actual in actuals.iter_mut().rev() {
            acc += *actual;
            *actual = acc;
        }

        (predictions, actuals)
    }

    macro_rules! run_down {
        ( $name:ident, $groups:tt, $size:tt ) => {
            #[test]
            fn $name() {
                let policy = &make_policy(&$groups);

                let (predictions, actuals) =
                        run_iterator(SeparatorIterator::new(policy, $size));

                assert_eq!(predictions, actuals);
            }
        };
    }

    run_down!(by_nothing_of_10, [], 10);

    run_down!(by_3s_of_10, [3], 10);

    run_down!(by_2s_of_10, [2], 10);
    run_down!(by_2s_of_9, [2], 9);
    run_down!(by_2s_of_8, [2], 8);
    run_down!(by_2s_of_7, [2], 7);
    run_down!(by_2s_of_6, [2], 6);
    run_down!(by_2s_of_5, [2], 5);
    run_down!(by_2s_of_4, [2], 4);
    run_down!(by_2s_of_3, [2], 3);
    run_down!(by_2s_of_2, [2], 2);
    run_down!(by_2s_of_1, [2], 1);
    run_down!(by_2s_of_0, [2], 0);

    run_down!(by_1s23_of_10, [3, 2, 1], 10);
    run_down!(by_1s23_of_9, [3, 2, 1], 9);
    run_down!(by_1s23_of_8, [3, 2, 1], 8);
    run_down!(by_1s23_of_7, [3, 2, 1], 7);
    run_down!(by_1s23_of_6, [3, 2, 1], 6);
    run_down!(by_1s23_of_5, [3, 2, 1], 5);
    run_down!(by_1s23_of_4, [3, 2, 1], 4);
    run_down!(by_1s23_of_3, [3, 2, 1], 3);
    run_down!(by_1s23_of_2, [3, 2, 1], 2);
    run_down!(by_1s23_of_1, [3, 2, 1], 1);
    run_down!(by_1s23_of_0, [3, 2, 1], 0);
}


src/policies.rs
use crate::digits::*;

/// A policy for inserting separators into numbers.
///
/// The configurable aspects are:
///
///   - The separator character to insert.
///
///   - How to group the separators.
///
///   - What characters are considered digits (for skipping non-digits such as
///     a minus sign).
#[derive(Debug, Clone, Copy)]
pub struct SeparatorPolicy<'a> {
    /// The separator to insert.
    pub separator: &'a str,
    /// The grouping. The numbers in this array give the size of the groups, from
    /// right to left, with the last number in the array giving the size of all
    /// subsequent groups.
    ///
    /// So to group by threes, as is typical in many places,
    /// this array should be `&[3]`. However, to get a grouping like `1,23,45,678`,
    /// where the last group has size three and the others size two, you would use
    /// `&[3, 2]`.
    pub groups:    &'a [u8],
    /// The characters that are considered digits. If there are multiple groups of
    /// digits separated by non-digits, we only add separators to the first group.
    /// This means, for example, that the number `-12345.67` will only have separators
    /// inserted into the `12345` portion.
    pub digits:    &'a [char],
}

/// Policy for placing a comma every three decimal digits.
pub const COMMA_SEPARATOR: SeparatorPolicy = SeparatorPolicy {
    separator:  ",",
    groups:     &[3],
    digits:     ASCII_DECIMAL,
};

/// Policy for placing a space every three decimal digits.
pub const SPACE_SEPARATOR: SeparatorPolicy = SeparatorPolicy {
    separator:  " ",
    groups:     &[3],
    digits:     ASCII_DECIMAL,
};

/// Policy for placing a period every three decimal digits.
pub const DOT_SEPARATOR: SeparatorPolicy = SeparatorPolicy {
    separator:  ".",
    groups:     &[3],
    digits:     ASCII_DECIMAL,
};

/// Policy for placing an underscore every three decimal digits.
pub const UNDERSCORE_SEPARATOR: SeparatorPolicy = SeparatorPolicy {
    separator:  "_",
    groups:     &[3],
    digits:     ASCII_DECIMAL,
};

/// Policy for placing a space every four hexadecimal digits.
pub const HEX_FOUR: SeparatorPolicy = SeparatorPolicy {
    separator:  " ",
    groups:     &[4],
    digits: ASCII_HEXADECIMAL,
};


src/lib.rs
#![doc(html_root_url = "https://docs.rs/thousands/0.1.4")]
//! Provides a trait, [`Separable`], for formatting numbers with
//! separators between the digits. Typically this will be used to add
//! commas or spaces every three digits from the right, but can be
//! configured via a [`SeparatorPolicy`].
//!
//! # Examples
//!
//! The simplest way to use the library is with trait [`Separable`]’s method
//! [`separate_with_commas`] method, which does what it sounds like:
//!
//! ```
//! use thousands::Separable;
//!
//! assert_eq!(   12345  .separate_with_commas(),  "12,345" );
//! assert_eq!( (-12345) .separate_with_commas(), "-12,345" );
//! assert_eq!(    9876.5.separate_with_commas(),   "9,876.5" );
//! ```
//!
//! There are also methods [`separate_with_spaces`], [`separate_with_dots`], and
//! [`separate_with_underscores`], in case you, your culture, or your file
//! format prefer those separators.
//!
//! However, it's also possible to pass a policy for different behavior:
//!
//! ```
//! use thousands::{Separable, SeparatorPolicy, digits};
//!
//! let policy = SeparatorPolicy {
//!     separator: ",",
//!     groups:    &[3, 2],
//!     digits:    digits::ASCII_DECIMAL,
//! };
//!
//! assert_eq!( 1234567890.separate_by_policy(policy), "1,23,45,67,890" );
//! ```
//!
//! # Usage
//!
//! It’s [on crates.io](https://crates.io/crates/thousands), so you can add
//!
//! ```toml
//! [dependencies]
//! thousands = "0.1.4"
//! ```
//!
//! to your `Cargo.toml`.
//!
//! This crate supports Rust version 1.22 and newer.
//!
//! [`Separable`]: trait.Separable.html
//! [`SeparatorPolicy`]: struct.SeparatorPolicy.html
//! [`separate_with_commas`]: trait.Separable.html#method.separate_with_commas
//! [`separate_with_spaces`]: trait.Separable.html#method.separate_with_spaces
//! [`separate_with_dots`]: trait.Separable.html#method.separate_with_dots
//! [`separate_with_underscores`]: trait.Separable.html#method.separate_with_underscores

/// Collections of digits.
///
/// These are used for constructing [SeparatorPolicy](struct.SeparatorPolicy.html)s.
pub mod digits;

/// Predefined policies.
pub mod policies;
pub use policies::SeparatorPolicy;

mod traits;
pub use traits::Separable;

// Contains blanket impl<T: Display> Separable for T.
mod display;

mod helpers;
